### ISSUE TITLE
Request help on the dataflow

### ISSUE THREAD

**Issue by Remember2018** [tensorpack/tensorpack#1366]
Hi Yuxin, I'm trying to port the `GroupBatchSampling` in `pytorch` to `Tensorpack` dataflow. And wrote the following code:

    class AspectGroupingDataFlow(RNGDataFlow):
        '''
        https://github.com/facebookresearch/maskrcnn-benchmark/blob/master/maskrcnn_benchmark/data/samplers/grouped_batch_sampler.py
        '''
    
        def __init__(self, roidbs, ds, group_ids, batch_size=2,  drop_uneven=True):
            super(AspectGroupingDataFlow, self).__init__()
            self.roidbs = roidbs
            self.ds = ds
            self.group_ids = group_ids
            self.batch_size = batch_size
            self.drop_uneven = drop_uneven
    
            self.groups = np.sort(np.unique(self.group_ids))
            self._can_reuse_batches = False
    
        def __iter__(self):
            if self._can_reuse_batches:
                batches = self._batches
                self._can_reuse_batches = False
            else:
                batches = self._prepare_batches()
            self._batches = batches
            for ids in iter(batches):
                batch_roidbs = [self.roidbs[idx] for idx in ids]
                yield batch_roidbs
    
        def get_data(self):
            return self.__iter__()
    
        def __len__(self):
            if not hasattr(self, '_batches'):
                self._batches = self._prepare_batches()
                self._can_reuse_batches = True
            return len(self._batches)
    
        def _prepare_batches(self):
            dataset_size = len(self.group_ids)
            sampled_ids = np.array(list(self.ds))
            order = np.full((dataset_size,), fill_value=-1, dtype=np.int64)
            order[sampled_ids] = np.arange(len(sampled_ids))
            mask = order >= 0
            clusters = [(self.group_ids == i) & mask for i in self.groups]
            relative_order = [order[cluster] for cluster in clusters]
            permutation_ids = [np.sort(s) for s in relative_order]
            permuted_clusters = [sampled_ids[idx] for idx in permutation_ids]
            # splits = [np.array_split(c, int(np.ceil(len(c)/self.batch_size))) for c in permuted_clusters]
            splits = [np.split(c, self.batch_size * np.arange(1, int(np.ceil(len(c) / self.batch_size))))
                      for c in permuted_clusters]
            merged = tuple(itertools.chain.from_iterable(splits))
            first_element_of_batch = [t[0] for t in merged]
            inv_sampled_ids_map = {v: k for k,v in enumerate(sampled_ids.tolist())}
            first_index_of_batch = [inv_sampled_ids_map[s] for s in first_element_of_batch]
            permutation_order = np.argsort(first_index_of_batch)
            batches = [merged[i].tolist() for i in permutation_order]
    
            if self.drop_uneven:
                kept = []
                for batch in batches:
                    if len(batch) == self.batch_size:
                        kept.append(batch)
                    batches = kept
            return batches
    
    
    def get_train_dataflow(batch_size=2):
        print("In train dataflow")
        roidbs = list(itertools.chain.from_iterable(DatasetRegistry.get(x).training_roidbs() for x in cfg.DATA.TRAIN))
        print_class_histogram(roidbs)
        print("Done loading roidbs")
    
        # Filter out images that have no gt boxes, but this filter shall not be applied for testing.
        # The model does support training with empty images, but it is not useful for COCO.
        num = len(roidbs)
        roidbs = list(filter(lambda img: len(img["boxes"][img["is_crowd"] == 0]) > 0, roidbs))
        logger.info(
            "Filtered {} images which contain no non-crowd groudtruth boxes. Total #images for training: {}".format(
                num - len(roidbs), len(roidbs)
            )
        )
    
        aspect_grouping = [1]
        aspect_ratios = [float(x["height"]) / float(x["width"]) for x in roidbs]
        group_ids = _quantize(aspect_ratios, aspect_grouping)
    
        ds = DataFromList(np.arange(len(roidbs)), shuffle=True)
        ds.reset_state()
        ds = AspectGroupingDataFlow(roidbs, ds, group_ids, batch_size=batch_size, drop_uneven=True)
        preprocess = TrainingDataPreprocessor()
        buffer_size = cfg.DATA.NUM_WORKERS * 20
        ds = MultiProcessMapData(ds, cfg.DATA.NUM_WORKERS, preprocess, buffer_size=buffer_size)
        ds.reset_state()
    
        # to get an infinite data flow
        ds = RepeatedData(ds, num=-1)
        dataiter = ds.__iter__()
    
        return dataiter

After this, the dataflow was used as follows:

    register_coco(os.path.expanduser(cfg.DATA.BASEDIR))

    data_iter = get_train_dataflow(batch_size=cfg.TRAIN.BATCH_SIZE_PER_GPU * num_gpus)
    ds = tf.data.Dataset()
    ds = ds.from_generator(lambda: map(lambda x: tuple([x[k] for k in
                                                        ['images', 'gt_boxes', 'gt_labels', 'orig_gt_counts',
                                                         'anchor_labels', 'anchor_boxes']]),
                                       data_iter),
                           (tf.float32, tf.float32, tf.int64, tf.int32, tf.int32, tf.float32),
                           (tf.TensorShape([None, None, None, 3]),
                            tf.TensorShape([None, None, 4]),
                            tf.TensorShape([None, None]),
                            tf.TensorShape([None, ]),
                            tf.TensorShape([None, None, None, None]),
                            tf.TensorShape([None, None, None, None, 4])))
    ds = ds.repeat()
    ds = ds.prefetch(buffer_size=256)
    ds = ds.make_one_shot_iterator()
    net_inputs = ds.get_next()

Then the `net_inputs` were directly used as inputs for model without `feed_dict`. There is a problem, the training is OK, but after some iterations (e.g., ~26000, with num_gpus=4, cfg.TRAIN.BATCH_SIZE_PER_GPU=4), the training is stuck. I guess the dataflow was ended after one epoch (the coco2017 has about 118000 images). I have set a `RepeatedData` in the dataflow, and the `ds = ds.repeat()` was also used, I cannot figure out what's wrong with the code. If convenient, could you please help give some advice? Thanks.

**Comment by ppwwyyxx** [comment]
If you think dataflow is the cause, you can just run it alone without training to see whether it will get stuck, and start from there.
One possible reason why such dataflow could get stuck is that it takes too much memory and some dataflow workers are killed by the operating system.

**Comment by Remember2018** [comment]
Good idea! I'll try that. In fact, I've tested the dataflow in small dataset `roidbs = roidbs[:5000]`, seems OK. I'm not sure if I understood the dataflow scheme correctly. Thank you Yuxin! @ppwwyyxx
